### PR TITLE
add tests to verify blitframebuffer when read pixels out-of-bound w/ and w/o filter

### PR DESCRIPTION
**Pull request by Richard-Yunchao**
@zhenyao and @kenrussell , PTAL. 

See the comments at https://bugs.chromium.org/p/chromium/issues/detail?id=644740

### CODE REVIEW

File: sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+             [-2, 4, 1, 7], // May not filter the image, dst region has the same width/height with src region.
+        ];
+
+        var readbufferHasSRGBImage = (readbufferFormat == gl.SRGB8_ALPHA8);
+        var drawbufferHasSRGBImage = (drawbufferFormat == gl.SRGB8_ALPHA8);
+
+        for (var i = 0; i < test.length; ++i) {
+            debug("");
+            switch (i) {
+                case 0: debug("Filter the image in read buffer: the width/height of dst region are different from width/height of src region."); break;
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html]: Why don't we put the message as part of the test case?

File: sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    gl.texImage2D(gl.TEXTURE_2D, 0, drawbufferFormat, size_draw, size_draw, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
+
+    var fbo_draw = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fbo_draw);
+    gl.framebufferTexture2D(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex_draw, 0);
+
+    if (gl.checkFramebufferStatus(gl.READ_FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE) {
+        // Blit read framebuffer to the image in draw framebuffer.
+        var test = [
+             [-2, 4, 1, 4], // Filter the image, dst region has different width/height with src region.
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html]: either add a comment or use dictionary so each param's semantic is clear.

File: sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  var tolerance = 3;
+  return (Math.abs(color[0] - expectedColor[0]) <= tolerance &&
+          Math.abs(color[1] - expectedColor[1]) <= tolerance &&
+          Math.abs(color[2] - expectedColor[2]) <= tolerance &&
+          Math.abs(color[3] - expectedColor[3]) <= tolerance);;
+}
+
+function blitframebuffer_filter_outofbounds(readbufferFormat, drawbufferFormat) {
+    debug("");
+    debug("bliting outside of read framebuffer, read buffer format is: " + wtu.glEnumToString(gl, readbufferFormat) + ", draw buffer format is: " + wtu.glEnumToString(gl, drawbufferFormat));
```
Comment (kenrussell) [inline sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html]: bliting -> blitting

File: sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    var color = 0x20;
+    for (var ii = 0; ii < size_read * size_read * 4; ii += 4) {
+        for (var jj = 0; jj < 3; ++jj) {
+          uint_read[ii + jj] = color;
+        }
+        uint_read[ii + 3] = 0xff;
+    }
+
+    // Create read framebuffer and feed data to read buffer
+    // Read buffer may has srgb image
```
Comment (kenrussell) [inline sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html]: may has -> may have (here and below)

File: sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+    var fbo_draw = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, fbo_draw);
+    gl.framebufferTexture2D(gl.DRAW_FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex_draw, 0);
+
+    if (gl.checkFramebufferStatus(gl.READ_FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE) {
+        // Blit read framebuffer to the image in draw framebuffer.
+        var test = [
+             [-2, 4, 1, 4], // Filter the image, dst region has different width/height with src region.
+             [-2, 4, 1, 7], // May not filter the image, dst region has the same width/height with src region.
```
Comment (kenrussell) [inline sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html]: For these two lines: with src -> as src

**Comment by zhenyao** [comment]
Per discussions from crbug.com/644740, the related deqp test cases are correct and we should not skip them.  With that considered, do we still need this standalone test?

File: sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        ];
+
+        var readbufferHasSRGBImage = (readbufferFormat == gl.SRGB8_ALPHA8);
+        var drawbufferHasSRGBImage = (drawbufferFormat == gl.SRGB8_ALPHA8);
+
+        for (var i = 0; i < test.length; ++i) {
+            debug("");
+            switch (i) {
+                case 0: debug("Filter the image in read buffer: the width/height of dst region are different from width/height of src region."); break;
+                case 1: debug("May not filter the image in read buffer: the width/height of dst region are the same with width/height of src region."); break;
```
Comment (kenrussell) [inline sdk/tests/conformance2/rendering/blitframebuffer-filter-outofbounds.html]: Could you be more precise about the differences between tests 0 and 1? Specifically, the behavior being tested here? I don't see where filtering of the image in the read buffer, or lack of filtering, is being verified. Could you point to specific areas in the OpenGL ES 3.0 specification?

**Comment by kenrussell** [comment]
If the dEQP tests are correct, then I would say no additional standalone test is needed.

**Comment by Richard-Yunchao** [comment]
Add more tests (NEAREST filter, dst pixels out-of-bounds, etc). Not for merge, but much easier to verify the out-of-bounds feature than deqp. @zhenyao  and @kenrussell

**Comment by zhenyao** [comment]
Do we have out-of-bounds blitFramebuffer tests with NEAREST in DEQP?  If not, we should consider merging this one.

**Comment by kenrussell** [comment]
Looking through deqp/functional/gles3/es3fFramebufferBlitTests.js, I don't see out-of-bounds blit tests with the NEAREST filter. Thanks @Richard-Yunchao for adding them.

At this point I am in favor of merging this pull request (squash and merge) to integrate these targeted tests.

**Comment by kenrussell** [comment]
@zhenyao what do you think?

**Comment by zhenyao** [comment]
Sounds good to me.  Merging now.

File: sdk/tests/conformance2/rendering/00_test_list.txt [hunk 7]
```diff
@@ -1,4 +1,5 @@
 attrib-type-match.html
+--min-version 2.0.1 blitframebuffer-filter-outofbounds.html
```
Comment (zhenyao) [inline sdk/tests/conformance2/rendering/00_test_list.txt]: Sorry, again, can we remove this min-version?  This should be part of 2.0.0 in my opinion.

**Comment by zhenyao** [comment]
One small issue, then I'll be happy to merge.

File: sdk/tests/conformance2/rendering/00_test_list.txt (hunk 7 above)
Comment (Richard-Yunchao) [inline sdk/tests/conformance2/rendering/00_test_list.txt]: done

**Pull request merged**